### ISSUE TITLE
Replace callback-based API with promise-based async functions

### ISSUE THREAD

**Issue by kibertoad** [lbdremy/solr-node-client#286]
Replace Request with superagent (which is faster: https://github.com/fdesjardins/node-http-client-bench) or axios (which is more popular) for promise support.

Undici is an even faster option (x2-x3 speed-up), but since it does not rely on `http` module, it would break nock-based mocks if users had any, which makes me hesitant.

**Comment by rudolfbyker** [comment]
x2-x3 speedup in what sense? The biggest time consumers are always sending / waiting for / receiving the network response.

**Comment by rudolfbyker** [comment]
I would prefer axios, because I favour maintainability. Having a well-known / popular package is better in that regard. I don't have much time left to work on this repo this month.

**Comment by rudolfbyker** [comment]
This issue replaces https://github.com/lbdremy/solr-node-client/issues/252

**Comment by rudolfbyker** [comment]
This issue replaces https://github.com/lbdremy/solr-node-client/issues/251 , too!

**Comment by kibertoad** [comment]
@rudolfbyker See https://www.youtube.com/watch?v=tr057sP4VBM for more details on Undici and overall deep-dive into HTTP client performance.
Well, superagent is also a well-supported and popular library, we are talking 4 million weekly downloads vs 15 millions weekly downloads, so both of them are very mature and popular.

See https://github.com/fdesjardins/node-http-client-bench for (somewhat outdated) benchmarks.

I would still prefer using Undici if we are ok with a breaking change that would force people to redo mocks with builtin Undici mocking, but I would need your input on that.

**Comment by rudolfbyker** [comment]
I'm not concerned with breaking changes. We are moving towards the first major release, so I think we should be relentless. If we had a big software house backing the project, that would be a different story.

I'll check out undici and get back to you. I need to get this done today.

**Comment by rudolfbyker** [comment]
OK, I see the advantage for situations where you have many small requests in a short time span. Like when doing autocomplete queries.

**Comment by rudolfbyker** [comment]
FYI https://github.com/nodejs/undici/issues/1011

**Comment by rudolfbyker** [comment]
- Undici does not support providing a custom HTTP agent. So we will have to drop that. (If I understand correctly from the, Undici IS an HTTP agent? But I'm confused about that.)

**Comment by kibertoad** [comment]
@rudolfbyker Yes, more or less. Since it does not use `http` Node module (it is more low-level than that), it does not support an `http` Agent either.

**Comment by rudolfbyker** [comment]
Let's also get rid of the custom error types and let Undici just do its thing. Extending errors in JS is problematic at best, and so we should not force this on users of our library.

**Comment by kibertoad** [comment]
Yeah, fair.

**Comment by rudolfbyker** [comment]
I'm a bit stuck on refactoring the `createAddStream` method. I'm not quite sure what it's supposed to do, or how to refactor the "duplexing".

**Comment by kibertoad** [comment]
@rudolfbyker It's a way to create a large amount of documents in bulk. Imagine that you have a billion of DB entries that you want indexes. What you might want to do is read them from DB as a stream and keep indexing them as you go.
I'm not sure you need to refactor much there, just replacing underlying Request call with Undici call should suffice. This is one case where still exposing a callback is a good idea, because that's streams are typically consumed.

**Comment by rudolfbyker** [comment]
I've never worked with streams in JS before, so I'm having a hard time wrapping my head around this. Should I be using undici.stream or unidici.pipeline?

**Comment by kibertoad** [comment]
@rudolfbyker I don't think you strictly _need_ to use either, original implementation just takes a plain simple Request call and wraps a Duplex (which writes to request and reads from response).
Now if you would use `undici.pipeline`, it should do all that job for you, `Dispatcher.pipeline` should create exactly the kind of Duplexer that we create ourselves now.
I would recommend against refactoring this part right now, though. There are no tests, so difficult to say if we break something in a major way. I suggest doing bare minimum change (replace simple Request call with simple Undici call) for now, and later on I will write a test and attempt a proper refactoring. Does that work for you?

**Comment by rudolfbyker** [comment]
The thumbs up was to say that I agree with the approach in principle, but I have been at this for a long time now, and can't get it to work. Do you have time to look into refactoring the `createAddStream` function specifically? It could even be a separate PR on its own, while everything else still uses `requests`.

**Comment by kibertoad** [comment]
Sure, I'll take a look. Can you push your branch?

**Comment by rudolfbyker** [comment]
https://github.com/lbdremy/solr-node-client/compare/issue-286-promises

**Comment by rudolfbyker** [comment]
I will have to stop work for today, as I'm starting to feel unwell. Thanks for the great collaboration!

**Comment by kibertoad** [comment]
@rudolfbyker Please feel better soon!
I'll try to take it over from here. Thank you, it's fun :)

**Comment by kibertoad** [comment]
@rudolfbyker I assume you didn't get simple Undici requests to work either yet?

**Comment by rudolfbyker** [comment]
I did. They are easier. but I did not push them, as they were only half
done. They were not completed yet because it causes a huge cascade of
refactoring everything to use promises, which I did not have time to
complete.

**Comment by rudolfbyker** [comment]
And thanks!

**Comment by kibertoad** [comment]
ok, I'll avoid touching those parts yet and focus only on streams

**Comment by rudolfbyker** [comment]
No worries. You can go ahead and refractor everything you want. I will
probably be able to work on this again in about two weeks.

**Comment by kibertoad** [comment]
can you push your half-done changes as well then?

**Comment by rudolfbyker** [comment]
I can do that later today. I'm not at my workstation right now.

**Comment by kibertoad** [comment]
thanks!

**Comment by rudolfbyker** [comment]
Sorry for the delay. Pushed now. See commit messages for details.

**Comment by rudolfbyker** [comment]
I'm going to take another stab at this, if you haven't started.

**Comment by kibertoad** [comment]
not yet :). thanks!

**Comment by rudolfbyker** [comment]
OK, that's all I have time for now.

**Comment by kibertoad** [comment]
Thank you! Hopefully will be able to continue the subject this week.

**Comment by kibertoad** [comment]
@rudolfbyker I will still add test for streaming functionality, and @AlinaKul will update the tests. Is there anything else left beyond that?

**Comment by rudolfbyker** [comment]
I cannot remember. You will have to dive into this. Not sure when I will have time to continue with this one.

**Comment by kibertoad** [comment]
@rudolfbyker Gotcha, will do! Thank you for the incredible work you've done on this!
